X3d

wangyang59 opened the pull request:
Initial draft of integrating xworld environments into flare. 
Please have a look at the general structure first. I will add tests later on.

yu239 commented on flare/framework/env.py:
```diff
@@ -1,25 +1,163 @@
-class GymEnv(object):
+class Env(object):
+    """
+    An abstract class for environment. A new environment should inherit this
```
inherit -> inherit from

yu239 commented on flare/framework/env.py:
```diff
@@ -1,25 +1,163 @@
+        init_ob = self.env.get_state()
+        screen, sentence = self.preprocess_observation(init_ob)
+
+        self.buf = [np.zeros(screen.shape).astype(screen.dtype) \
```
xworld by default has an option of 'context'. So probably you don't need to maintain here.

yu239 commented on flare/framework/env.py:
```diff
@@ -1,25 +1,163 @@
+        assert contexts >= 1, "contexts must be a positive number!"
+        self.contexts = contexts
+        self.buf = None
+        self.dict_id2w = dict
```
Only a list is enough for converting id to word

yu239 commented on flare/framework/env.py:
```diff
@@ -1,25 +1,163 @@
+        next_game_over = self.env.game_over() != "alive"
+        next_ob = self.env.get_state()
+
+        screen, sentence = self.preprocess_observation(next_ob)
```
If the option 'context' is set, you don't have to maintain a buffer.

yu239 commented on flare/framework/env.py:
```diff
@@ -49,6 +187,7 @@ def step(self, actions, actrep=1):
         total_reward = 0
         next_game_over = False
         for i in range(actrep):
+            self.steps += 1
```
This should be moved outside of the for loop.

yu239 commented on flare/framework/env.py:
```diff
@@ -1,25 +1,163 @@
     """
 
-    def __init__(self, game_name, contexts=1):
+    def __init__(self, game_name, contexts=1, options=None):
```
delete 'options=None'

yu239 commented on flare/framework/env.py:
```diff
@@ -1,25 +1,163 @@
+    @abstractmethod
+    def reset(self):
+        """
+        reset the environment and return the initial observation
```
return a list of initial observations

yu239 commented on flare/framework/env.py:
```diff
@@ -1,25 +1,163 @@
 import numpy as np
 from scipy import misc
+from abc import ABCMeta, abstractmethod
+from py_simulator import Simulator
```
This might cause errors if the user hasn't installed py_simulator yet. Probably we should put the import inside the class?

wangyang59 commented:
@yu239 I have modified the codes according to your initial reviews. I have moved gym_env and xworld env into a separate folder of `env_zoo` and added a test for xworld_env

yu239 commented on flare/framework/env.py:
```diff
@@ -45,10 +41,9 @@ def observation_dims(self):
-        For each integer,
-        if the corresponding action is discrete, then it means
-        the total number of actions;
+        Return a tuple of integers as action dimensions, each integer for an 
```
tuple -> list. Use list for elements of different kinds

wangyang59 commented:
@yu239 I have rebased and changed the code accordingly.
